File: packages/sdk-ts/src/kernel/hooks.ts
/**
 * HookRegistry — middleware (use) + events (on / onAsync) for kernel lifecycle.
 *
 * 1:1 parity with Python dna.v3.kernel.hooks.
 *
 * Dual-channel events:
 *   - Sync listeners (legacy + in-process metrics/logging):
 *       k.on("post_build_prompt", fn)        // fn: (ctx) => void
 *       k.hooks.emit("post_build_prompt", ctx)
 *   - Async listeners (I/O-heavy: event bus publishers, HTTP, DB writes):
 *       k.onAsync("post_save", fn)            // fn: (ctx) => Promise<void>
 *       await k.hooks.emitAsync("post_save", ctx)
 *
 * Sync `on()` autodetects based on whether the handler returns a Promise.
 * Callers that invoke `emit` (sync) skip async listeners — callers that
 * invoke `emitAsync` run both channels. The skip is LOUD
 * (s-kernel-fail-soft-audit, Py twin semantics): `emit` counts every skip
 * in `skippedAsyncEmits` and console.warns ONCE per (hook, listener);
 * `emit(hook, ctx, { strict: true })` throws instead — for call sites
 * where skipping a listener would be a bug, not lost telemetry. The
 * kernel's own sync emits (parse_error / kinddef_conflict /
 * extension_error / sync-path post_build_prompt) are observability
 * funnels and stay non-strict.
 *
 * Veto hooks (s-write-path-despecialize): ordered listeners that can BLOCK
 * the operation by throwing. Unlike `emit`/`emitAsync` (fire-and-forget,
 * errors swallowed), `emitVeto` runs listeners in ascending `priority`
 * order and lets exceptions PROPAGATE — a throw IS the veto. Listeners may
 * also mutate the typed context in place (e.g. fields on `ctx.raw`) — the
 * write proceeds with the mutated payload.
 *
 *   k.hooks.onVeto("pre_save", fn, { priority: 20, key: "helix.rule" });
 *   await k.hooks.emitVeto("pre_save", ctx);
 *
 * `key` makes registration idempotent: re-registering the same key REPLACES
 * the previous listener (extensions may be re-loaded onto a shared registry).
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * The full vocabulary of hook names the SDK itself emits/consumes
 * (s-dna-typed-hook-names). Py twin: `HookName` / `KNOWN_HOOK_NAMES` in
 * dna/kernel/hooks.py. Both sides are locked to the shared fixture
 * `tests/parity-fixtures/port-surface-parity.json` (section `hook_names`)
 * by tests/hook-names.test.ts + tests/test_hook_names.py — vocabulary
 * drift turns both suites red. Add a name HERE + fixture + Py, never ad hoc.
 */
export const KNOWN_HOOK_NAMES = [
  "pre_build_prompt",
  "post_build_prompt",
  "pre_save",
  "post_save",
  "post_delete",
  "kinddef_conflict",
  "parse_error",
  "extension_error",
] as const;

export type HookName = (typeof KNOWN_HOOK_NAMES)[number];

/**
 * Parameter type for hook-name arguments: the known vocabulary (with editor
 * autocomplete) plus arbitrary strings for back-compat (custom hooks stay
 * legal — they warn at runtime, never break). A misspelled builtin name
 * like `on("pre_saev", fn)` used to compile, run, and never fire, silently;
 * the registry now console.warns once per (registry, name).
 */
export type HookNameArg = HookName | (string & {});

const KNOWN_HOOK_SET: ReadonlySet<string> = new Set(KNOWN_HOOK_NAMES);

export interface HookContext {
  scope: string;
  agent?: string;
  kind?: string;
  name?: string;
  data: Record<string, unknown>;
  prompt?: string;
  /** Optional layer tuple when the write targeted a layer overlay.
   *  `undefined` means the write went to the scope's base. Set by
   *  Kernel.writeDocument when a layer kwarg is passed (Phase 2a.0). */
  layer?: [string, string];
}

export type Middleware = (ctx: HookContext) => HookContext;
export type EventHandler = (ctx: HookContext) => void;
export type AsyncEventHandler = (ctx: HookContext) => Promise<void>;

/** Typed context for the `pre_save` veto hook (s-write-path-despecialize).
 *  Handed to every veto listener BEFORE the document reaches the source
 *  adapter. `raw` is the live payload — listeners may mutate it in place.
 *  `kernel` is the kernel the write flows through (tenant-bound copy when
 *  applicable) so guards can read (getDocument, ...). Throwing vetoes the
 *  write. Py twin: dna.kernel.hooks.PreSaveContext. */
export interface PreSaveContext {
  scope: string;
  kind: string;
  name: string;
  raw: Record<string, unknown>;
  tenant: string | null;
  layer?: [string, string];
  kernel: unknown;
}

/** Veto listener: sync or async; throwing vetoes the operation. */
export type VetoHandler = (ctx: PreSaveContext) => void | Promise<void>;

// ---------------------------------------------------------------------------
// Registry
// ---------------------------------------------------------------------------

export class HookRegistry {
  private _middleware: Map<string, Middleware[]> = new Map();
  private _events: Map<string, EventHandler[]> = new Map();
  private _asyncEvents: Map<string, AsyncEventHandler[]> = new Map();
  /** Veto channel — hook → list of {priority, seq, key, fn}. `seq` keeps
   *  registration order stable among equal priorities. */
  private _veto: Map<
    string,
    Array<{ priority: number; seq: number; key: string | null; fn: VetoHandler }>
  > = new Map();
  private _vetoSeq = 0;
  /** s-kernel-fail-soft-audit — sync-emit async-listener skip tracking:
   *  warn ONCE per (hook, listener) + count every skip per hook so a
   *  mis-wired listener is visible instead of an invisible no-op. */
  private _asyncSkipWarned: Set<string> = new Set();
  readonly skippedAsyncEmits: Map<string, number> = new Map();
  /** s-dna-typed-hook-names — unknown-name warn-once tracking (per
   *  registry, per name; same posture as _asyncSkipWarned). */
  private _unknownHookWarned: Set<string> = new Set();

  /** Warn (once per registry+name) on a name outside `HookName`.
   *  Fail-loud, NOT fail-closed: custom hook names keep working
   *  (back-compat / third-party channels), but a typo of a builtin name —
   *  the listener that never fires — is now visible. Py twin semantics:
   *  `UnknownHookNameWarning`. */
  private _checkHookName(hook: string): void {
    if (KNOWN_HOOK_SET.has(hook) || this._unknownHookWarned.has(hook)) return;
    this._unknownHookWarned.add(hook);
    console.warn(
      `unknown hook name ${JSON.stringify(hook)} — known: ` +
      `${JSON.stringify([...KNOWN_HOOK_NAMES].sort())}. A listener on (or ` +
      `emit of) a misspelled hook never fires; custom names still work ` +
      `(warning only).`,
    );
  }

  /** Register middleware. Called in order, each receives output of previous. */
  use(hook: HookNameArg, fn: Middleware): void {
    this._checkHookName(hook);
    const list = this._middleware.get(hook) ?? [];
    list.push(fn);
    this._middleware.set(hook, list);
  }

  /**
   * Register event subscriber. Fire-and-forget, errors logged but not raised.
   *
   * Autodetects sync vs async via `fn.constructor.name === "AsyncFunction"`.
   * Async functions are routed to the async channel — callers using `emit`
   * (sync) skip them; `emitAsync` runs both.
   */
  on(hook: HookNameArg, fn: EventHandler | AsyncEventHandler): void {
    this._checkHookName(hook);
    // AsyncFunction detection: works for `async function`. For regular
    // functions that happen to return Promises, prefer `onAsync` explicitly.
    if ((fn as Function).constructor?.name === "AsyncFunction") {
      const list = this._asyncEvents.get(hook) ?? [];
      list.push(fn as AsyncEventHandler);
      this._asyncEvents.set(hook, list);
    } else {
      const list = this._events.get(hook) ?? [];
      list.push(fn as EventHandler);
      this._events.set(hook, list);
    }
  }

  /** Explicit async-only registration. Prefer `on()` — this is here for
   *  symmetry and for non-`async function` callables (arrow with Promise
   *  body, partials) where AsyncFunction detection returns false. */
  onAsync(hook: HookNameArg, fn: AsyncEventHandler): void {
    this._checkHookName(hook);
    const list = this._asyncEvents.get(hook) ?? [];
    list.push(fn);
    this._asyncEvents.set(hook, list);
  }

  /** Run middleware chain. Returns modified context. */
  runMiddleware(hook: HookNameArg, ctx: HookContext): HookContext {
    this._checkHookName(hook);
    for (const fn of this._middleware.get(hook) ?? []) {
      ctx = fn(ctx);
    }
    return ctx;
  }

  /**
   * Sync emit. Runs only sync listeners. If async listeners exist for this
   * hook they CANNOT run here — the caller must use `emitAsync` to reach
   * them. The skip is loud (s-kernel-fail-soft-audit):
   *
   * - default: each skipped listener is counted in `skippedAsyncEmits` and
   *   a console.warn fires once per (hook, listener) — a mis-wired async
   *   listener is visible, not an invisible no-op.
   * - `strict: true`: throw instead — for emit sites where skipping a
   *   listener would be a bug, not lost telemetry.
   */
  emit(hook: HookNameArg, ctx: HookContext, opts?: { strict?: boolean }): void {
    this._checkHookName(hook);
    for (const fn of this._events.get(hook) ?? []) {
      try {
        fn(ctx);
      } catch (e) {
        console.warn(`Hook ${hook} subscriber error:`, e);
      }
    }
    const skipped = this._asyncEvents.get(hook) ?? [];
    if (skipped.length > 0) {
      const total = (this.skippedAsyncEmits.get(hook) ?? 0) + skipped.length;
      this.skippedAsyncEmits.set(hook, total);
      if (opts?.strict) {
        const names = skipped.map((fn) => fn.name || "<anonymous>");
        throw new Error(
          `Hook ${JSON.stringify(hook)} was emitted via sync emit(strict) ` +
          `but has ${skipped.length} async listener(s) ` +
          `${JSON.stringify(names)} that cannot run — use emitAsync() at ` +
          `this call site.`,
        );
      }
      for (const fn of skipped) {
        const key = `${hook}\0${fn.name || "<anonymous>"}`;
        if (this._asyncSkipWarned.has(key)) continue;
        this._asyncSkipWarned.add(key);
        console.warn(
          `Hook ${hook} has async listener ${fn.name || "<anonymous>"} but ` +
          `was fired via sync emit(); the listener was SKIPPED (total ` +
          `skips for this hook: ${total}). Use emitAsync() to reach it. ` +
          `(warned once per (hook, listener))`,
        );
      }
    }
  }

  /**
   * Async emit. Runs sync listeners inline, then awaits each async listener.
   * Use from async call sites (kernel.writeDocument, kernel.deleteDocument).
   * Errors in any listener are logged and do NOT prevent other listeners.
   */
  async emitAsync(hook: HookNameArg, ctx: HookContext): Promise<void> {
    this._checkHookName(hook);
    for (const fn of this._events.get(hook) ?? []) {
      try {
        fn(ctx);
      } catch (e) {
        console.warn(`Hook ${hook} (sync) subscriber error:`, e);
      }
    }
    for (const fn of this._asyncEvents.get(hook) ?? []) {
      try {
        await fn(ctx);
      } catch (e) {
        console.warn(`Hook ${hook} (async) subscriber error:`, e);
      }
    }
  }

  // -- Veto channel (s-write-path-despecialize) ------------------------------

  /**
   * Register a veto listener on `hook`. Listeners run in ascending
   * `priority` (ties keep registration order). Throwing from a listener
   * PROPAGATES to the emitter — that is the veto. Sync and async callables
   * both work.
   *
   * `key` (recommended: `"<extension>.<rule>"`) makes registration
   * idempotent — a second `onVeto` with the same key REPLACES the earlier
   * listener instead of stacking a duplicate.
   */
  onVeto(
    hook: HookNameArg,
    fn: VetoHandler,
    opts?: { priority?: number; key?: string },
  ): void {
    this._checkHookName(hook);
    const listeners = this._veto.get(hook) ?? [];
    const key = opts?.key ?? null;
    const filtered = key !== null
      ? listeners.filter((entry) => entry.key !== key)
      : listeners;
    this._vetoSeq += 1;
    filtered.push({
      priority: opts?.priority ?? 0, seq: this._vetoSeq, key, fn,
    });
    filtered.sort((a, b) => a.priority - b.priority || a.seq - b.seq);
    this._veto.set(hook, filtered);
  }

  /**
   * Run veto listeners for `hook` in priority order. Unlike
   * `emit`/`emitAsync`, exceptions are NOT swallowed — the first throw
   * aborts the chain and propagates to the caller (the operation is
   * vetoed). Listeners may mutate `ctx` in place.
   */
  async emitVeto(hook: HookNameArg, ctx: PreSaveContext): Promise<void> {
    this._checkHookName(hook);
    for (const entry of this._veto.get(hook) ?? []) {
      await entry.fn(ctx);
    }
  }

  /** Check if any veto listeners are registered for `hook`. */
  hasVeto(hook: string): boolean {
    return (this._veto.get(hook)?.length ?? 0) > 0;
  }

  /** Check if any middleware, events (sync or async) or veto listeners are
   *  registered. */
  has(hook: string): boolean {
    return (
      (this._middleware.get(hook)?.length ?? 0) > 0 ||
      (this._events.get(hook)?.length ?? 0) > 0 ||
      (this._asyncEvents.get(hook)?.length ?? 0) > 0 ||
      (this._veto.get(hook)?.length ?? 0) > 0
    );
  }
}

File: packages/sdk-py/dna/kernel/hooks.py
"""HookRegistry — middleware (use) + events (on / on_async) for kernel lifecycle.

Middleware: intercepts and can modify data flowing through the kernel.
  k.use("pre_build_prompt", fn)  → fn(ctx) → modified ctx

Events: fire-and-forget notifications. Two channels:
  - Sync listeners (legacy + in-process metrics/logging):
      k.on("post_build_prompt", fn)     # fn: (ctx) -> None
      k.hooks.emit("post_build_prompt", ctx)
  - Async listeners (I/O-heavy: event bus publishers, network calls, db writes):
      k.on_async("post_save", fn)        # fn: (ctx) -> Awaitable[None]
      await k.hooks.emit_async("post_save", ctx)

Design rationale — dual channel over "make everything async":
  - Sync consumers (logging, metrics, error counters) don't need a loop and
    historically attach to hooks that fire in sync code paths (parse_error,
    post_build_prompt from a sync prompt_builder.build()). Forcing async on
    them would cascade-break the whole SDK.
  - Async consumers want to `await` directly (event bus publish, HTTP, DB)
    without the `asyncio.create_task` workaround that drops errors and has
    no backpressure.
  - Callers of `emit_async` (write_document, delete_document — already async)
    see both channels fire; sync listeners still run, async ones are awaited.
  - Callers of `emit` (sync code paths) run only sync listeners. Registering
    an async listener for a hook only fired from sync emit used to be an
    INVISIBLE no-op (debug-only log). Since s-kernel-fail-soft-audit the
    skip is loud: `emit` warns ONCE per (hook, listener) at WARNING level,
    counts every skip in `skipped_async_emits[hook]`, and accepts
    `strict=True` to raise instead — for emit sites where reaching every
    listener is load-bearing.

  Kernel-internal emit posture (s-kernel-fail-soft-audit, documented
  decision): the kernel's own sync emits (`parse_error`,
  `kinddef_conflict`, `extension_error`, and `post_build_prompt` from the
  SYNC prompt_builder path) are observability funnels — a mis-wired async
  listener there means lost telemetry, not corrupted state, so they emit
  with strict=False and rely on the warn-once. Async call sites reach both
  channels instead: `prompt_kernel.build_prompt_async` awaits `emit_async`
  for `post_build_prompt` (it used to fire sync `emit` and silently skip
  async listeners despite being in an async context).

Veto hooks (s-write-path-despecialize): ordered listeners that can BLOCK the
operation by raising. Unlike `emit`/`emit_async` (fire-and-forget, errors
swallowed), `emit_veto` runs listeners in ascending `priority` order and lets
exceptions PROPAGATE — a raise IS the veto. Listeners may also mutate the
typed context in place (e.g. preserve fields on `ctx.raw`) — the write
proceeds with the mutated payload. Sync and async listeners share one
ordered chain.

  k.hooks.on_veto("pre_save", fn, priority=20, key="helix.prompt-budget")
  await k.hooks.emit_veto("pre_save", PreSaveContext(...))

`key` makes registration idempotent: re-registering the same key REPLACES
the previous listener (extensions may be re-loaded onto a shared registry).

Hook points:
  - pre_build_prompt:  called before prompt rendering, can modify context
  - post_build_prompt: called after prompt rendering, receives final prompt
  - pre_save:          called before document save, can reject (veto channel;
                       emitted by kernel.write_document with PreSaveContext)
  - post_save:         called after document save  [fires via emit_async]
  - post_delete:       called after document delete [fires via emit_async]
"""
from __future__ import annotations

import inspect
import logging
from dataclasses import dataclass, field
from typing import Any, Awaitable, Callable, Union

logger = logging.getLogger(__name__)


@dataclass
class HookContext:
    """Data passed to middleware and event hooks."""
    scope: str = ""
    agent: str | None = None
    kind: str | None = None
    name: str | None = None
    data: dict[str, Any] = field(default_factory=dict)
    prompt: str | None = None
    layer: tuple[str, str] | None = None  # Layer context for post_save/post_delete (phase 2a.0)


@dataclass
class PreSaveContext:
    """Typed context for the ``pre_save`` veto hook.

    Handed to every veto listener BEFORE the document reaches the source
    adapter. ``raw`` is the live payload — listeners may mutate it in place
    (the write persists the mutated dict). ``kernel`` is the kernel the
    write is flowing through (tenant-bound copy when applicable) so guards
    can do reads (``get_document``, ``model_profile``, ...) without module
    globals. Raising from a listener vetoes the write.
    """
    scope: str
    kind: str
    name: str
    raw: dict[str, Any]
    tenant: str | None = None
    layer: tuple[str, str] | None = None
    kernel: Any = None


# Middleware: (HookContext) -> HookContext
MiddlewareFn = Callable[[HookContext], HookContext]
# Event (sync): (HookContext) -> None
EventFn = Callable[[HookContext], None]
# Event (async): (HookContext) -> Awaitable[None]
AsyncEventFn = Callable[[HookContext], Awaitable[None]]
# Veto listener: sync or async; raising vetoes the operation.
VetoFn = Callable[[Any], Union[None, Awaitable[None]]]


class HookRegistry:
    """Manages middleware chains and event subscribers (sync + async)."""

    def __init__(self) -> None:
        self._middleware: dict[str, list[MiddlewareFn]] = {}
        self._events: dict[str, list[EventFn]] = {}
        self._async_events: dict[str, list[AsyncEventFn]] = {}
        # Veto channel — hook → list of (priority, seq, key, fn). seq keeps
        # registration order stable among equal priorities.
        self._veto: dict[str, list[tuple[int, int, "str | None", VetoFn]]] = {}
        self._veto_seq: int = 0
        # s-kernel-fail-soft-audit — sync-emit async-listener skip tracking:
        # warn ONCE per (hook, listener) + count every skip per hook so a
        # mis-wired listener is visible instead of an invisible no-op.
        self._async_skip_warned: set[tuple[str, str]] = set()
        self.skipped_async_emits: dict[str, int] = {}

    def use(self, hook: str, fn: MiddlewareFn) -> None:
        """Register middleware. Called in order, each receives output of previous."""
        self._middleware.setdefault(hook, []).append(fn)

    def on(self, hook: str, fn: Union[EventFn, AsyncEventFn]) -> None:
        """Register event subscriber. Fire-and-forget, errors logged but not raised.

        Accepts either sync ``fn(ctx) -> None`` or async ``async fn(ctx) -> None``.
        Async functions are routed to the async channel automatically — callers
        that invoke ``emit`` (sync) will skip them; callers that invoke
        ``emit_async`` run both channels.
        """
        if inspect.iscoroutinefunction(fn):
            self._async_events.setdefault(hook, []).append(fn)  # type: ignore[arg-type]
        else:
            self._events.setdefault(hook, []).append(fn)  # type: ignore[arg-type]

    def on_async(self, hook: str, fn: AsyncEventFn) -> None:
        """Explicit async-only registration.

        Prefer ``on()`` — it autodetects. ``on_async`` is here for symmetry
        and for lambdas / partials where ``iscoroutinefunction`` returns False
        incorrectly.
        """
        self._async_events.setdefault(hook, []).append(fn)

    def run_middleware(self, hook: str, ctx: HookContext) -> HookContext:
        """Run middleware chain. Returns modified context."""
        for fn in self._middleware.get(hook, []):
            ctx = fn(ctx)
        return ctx

    def emit(self, hook: str, ctx: HookContext, *, strict: bool = False) -> None:
        """Sync emit. Runs only sync listeners.

        If async listeners exist for this hook they CANNOT run here — the
        caller must use ``emit_async`` to reach them. The skip is loud
        (s-kernel-fail-soft-audit):

        - ``strict=False`` (default): each skipped listener is counted in
          ``skipped_async_emits[hook]`` and a WARNING fires once per
          (hook, listener) — a mis-wired async listener is visible, not
          an invisible no-op. Used by the kernel's own observability
          funnels (parse_error / kinddef_conflict / extension_error /
          sync-path post_build_prompt) where taking the operation down
          over a listener wiring mistake would be worse than the skip.
        - ``strict=True``: raise RuntimeError instead — for emit sites
          where skipping a listener would be a bug, not lost telemetry.
        """
        for fn in self._events.get(hook, []):
            try:
                fn(ctx)
            except Exception as e:
                logger.warning("Hook %s subscriber error: %s", hook, e)
        skipped = self._async_events.get(hook)
        if skipped:
            self.skipped_async_emits[hook] = (
                self.skipped_async_emits.get(hook, 0) + len(skipped)
            )
            if strict:
                names = [getattr(fn, "__qualname__", repr(fn)) for fn in skipped]
                raise RuntimeError(
                    f"Hook {hook!r} was emitted via sync emit(strict=True) but "
                    f"has {len(skipped)} async listener(s) {names} that cannot "
                    f"run — use emit_async() at this call site."
                )
            for fn in skipped:
                key = (hook, getattr(fn, "__qualname__", repr(fn)))
                if key in self._async_skip_warned:
                    continue
                self._async_skip_warned.add(key)
                logger.warning(
                    "Hook %s has async listener %s but was fired via sync "
                    "emit(); the listener was SKIPPED (total skips for this "
                    "hook: %d). Use emit_async() to reach it. "
                    "(warned once per (hook, listener))",
                    hook, key[1], self.skipped_async_emits[hook],
                )

    async def emit_async(self, hook: str, ctx: HookContext) -> None:
        """Async emit. Runs sync listeners inline, then awaits each async listener.

        Use from async call sites (kernel.write_document, kernel.delete_document).
        Errors in any listener are logged and do NOT prevent other listeners
        from running.
        """
        # Sync listeners first (same behavior as emit()).
        for fn in self._events.get(hook, []):
            try:
                fn(ctx)
            except Exception as e:
                logger.warning("Hook %s (sync) subscriber error: %s", hook, e)
        # Async listeners — await each. No gather() so errors in one don't
        # cancel the others; simple linear await keeps ordering deterministic.
        for fn in self._async_events.get(hook, []):
            try:
                await fn(ctx)
            except Exception as e:
                logger.warning("Hook %s (async) subscriber error: %s", hook, e)

    # -- Veto channel (s-write-path-despecialize) -----------------------------

    def on_veto(
        self, hook: str, fn: VetoFn, *,
        priority: int = 0, key: str | None = None,
    ) -> None:
        """Register a veto listener on ``hook``.

        Listeners run in ascending ``priority`` (ties keep registration
        order). Raising from a listener PROPAGATES to the emitter — that is
        the veto. Sync and async callables both work.

        ``key`` (recommended: ``"<extension>.<rule>"``) makes registration
        idempotent — a second ``on_veto`` with the same key REPLACES the
        earlier listener instead of stacking a duplicate.
        """
        listeners = self._veto.setdefault(hook, [])
        if key is not None:
            listeners[:] = [entry for entry in listeners if entry[2] != key]
        self._veto_seq += 1
        listeners.append((priority, self._veto_seq, key, fn))
        listeners.sort(key=lambda entry: (entry[0], entry[1]))

    async def emit_veto(self, hook: str, ctx: Any) -> None:
        """Run veto listeners for ``hook`` in priority order.

        Unlike ``emit``/``emit_async``, exceptions are NOT swallowed — the
        first raise aborts the chain and propagates to the caller (the
        operation is vetoed). Listeners may mutate ``ctx`` in place.
        """
        for _prio, _seq, _key, fn in self._veto.get(hook, []):
            result = fn(ctx)
            if inspect.isawaitable(result):
                await result

    def has_veto(self, hook: str) -> bool:
        """Check if any veto listeners are registered for ``hook``."""
        return bool(self._veto.get(hook))

    def has(self, hook: str) -> bool:
        """Check if any middleware, events (sync or async) or veto listeners
        are registered."""
        return (
            bool(self._middleware.get(hook))
            or bool(self._events.get(hook))
            or bool(self._async_events.get(hook))
            or bool(self._veto.get(hook))
        )
